Wrappers/Python/cil/optimisation/functions/SVRGFunction.py

```diff
@@ -0,0 +1,275 @@
+                self._list_stored_gradients = [
+                    fi.gradient(x) for fi in self.functions]
+                self._full_gradient_at_snapshot = sum(self._list_stored_gradients, start=0*x)
+            else:
+                for i, fi in enumerate(self.functions):
+                    fi.gradient(x, out=self._list_stored_gradients[i])
+                self._full_gradient_at_snapshot.fill( sum(self._list_stored_gradients, start=0*x))
+           
+        else:
+            # Save the snapshot point and the gradient of the full sum at the point x. 
+            self._full_gradient_at_snapshot = self.full_gradient(x, out=self._full_gradient_at_snapshot) 
+            
+        if self.snapshot is None: 
+            self.snapshot = x.copy()
+
+        self.snapshot.fill(x) 
+
+        # In this iteration all functions in the sum were used to update the gradient 
+        self._update_data_passes_indices(list(range(self.num_functions))) 
+
+        # Return the gradient of the full sum at the snapshot. 
+        if out is None:
+            out = self._full_gradient_at_snapshot
+        else:
+            out.fill( self._full_gradient_at_snapshot)
+
+        return out
+
+
+class LSVRGFunction(SVRGFunction):
+    """""
+    A class representing a function for Loopless Stochastic Variance Reduced Gradient (SVRG) approximation. This is similar to SVRG, except the full gradient at a "snapshot"  is calculated at random intervals rather than at fixed numbers of iterations. 
+    
+    
+    Reference
+    ----------
+    
+    Kovalev, D., Horváth, S. &; Richtárik, P.. (2020). Don’t Jump Through Hoops and Remove Those Loops:  SVRG and Katyusha are Better Without the Outer Loop. Proceedings of the 31st International Conference  on Algorithmic Learning Theory, in Proceedings of Machine Learning Research 117:451-467 Available from https://proceedings.mlr.press/v117/kovalev20a.html.
+
+
+
+    Parameters
+    ----------
+     functions : `list`  of functions
+        A list of functions: :code:`[f_{0}, f_{1}, ..., f_{n-1}]`. Each function is assumed to be smooth with an implemented :func:`~Function.gradient` method. All functions must have the same domain. The number of functions `n` must be strictly greater than 1. 
+    sampler: An instance of a CIL Sampler class ( :meth:`~optimisation.utilities.sampler`) or of another class which has a `next` function implemented to output integers in {0,...,n-1}.
+        This sampler is called each time gradient is called and  sets the internal `function_num` passed to the `approximate_gradient` function.  Default is `Sampler.random_with_replacement(len(functions))`. 
+    update_prob: positive float, default: 1/n
+        The probability of updating the full gradient (taking a snapshot) at each iteration. The default is :math:`1./n` so, in expectation, a snapshot will be taken every :math:`n` iterations. 
```

Comment:
To mirror what suggested in SVRG, we could rename `update_prob` to `snapshot_update_probability`